pythonflow/pfmq.py

```diff
@@ -0,0 +1,399 @@
+                    LOGGER.debug('received cancel signal on %s', self._cancel_address)
+                    break
+
+        LOGGER.debug('exiting communication loop')
+
+    @classmethod
+    def from_graph(cls, graph, *args, **kwargs):
+        """
+        Construct a worker from a computational graph.
+
+        Parameters
+        ----------
+        graph : Graph
+            A computational graph.
+        """
+        def _target(request):
+            if 'context' in request:
+                return graph(request['fetches'], request['context'])
+            elif 'contexts' in request:
+                return [graph(request['fetches'], context) for context in request['contexts']]
+            raise KeyError
+        return cls(_target, *args, **kwargs)
+
+
+class MessageBroker(Base):
+    """
+    Message broker for tasks that are executed remotely.
+
+    Parameters
+    ----------
+    backend_address : str
+        Address to which workers can connect.
+    frontend_address : str
+        Address to which tasks can connect (defaults to a random in-process communication channel).
+    start : bool
+        Whether to start the event loop as a background thread.
+    """
+    def __init__(self, backend_address, frontend_address=None, start=False):
+        self.backend_address = backend_address
+        self.frontend_address = frontend_address or f'inproc://{uuid.uuid4().hex}'
+        super(MessageBroker, self).__init__(start)
+
+    def run(self):  # pylint: disable=too-many-statements,too-many-locals
+        context = zmq.Context.instance()
+        workers = set()
+        cache = {}
+
+        # pylint: disable=E1101
+        with context.socket(zmq.ROUTER) as frontend, context.socket(zmq.ROUTER) as backend, \
+            context.socket(zmq.PAIR) as cancel:
+        # pylint: enable=E1101
+
+            cancel.connect(self._cancel_address)
+            frontend.bind(self.frontend_address)
+            LOGGER.debug('bound frontend to %s', self.frontend_address)
+            backend.bind(self.backend_address)
+            LOGGER.debug('bound backend to %s', self.backend_address)
+
+            backend_poller = zmq.Poller()
+            backend_poller.register(backend, zmq.POLLIN)
+            backend_poller.register(cancel, zmq.POLLIN)
+
+            poller = zmq.Poller()
+            poller.register(frontend, zmq.POLLIN)
+            poller.register(backend, zmq.POLLIN)
+            poller.register(cancel, zmq.POLLIN)
+
+            while True:
+                if workers:
+                    LOGGER.debug('polling frontend and backend...')
+                    sockets = dict(poller.poll())
+                else:
+                    LOGGER.debug('polling backend...')
+                    sockets = dict(backend_poller.poll())
+
+                # Receive results or sign-up messages from the backend
+                if sockets.get(backend) == zmq.POLLIN:
+                    worker, _, client, *message = backend.recv_multipart()
```

Comment:
Could you explain where the `worker, _` header comes from? In the `Worker` above, I can only see `socket.send_multipart([client, _, identifier, result])`.